File: entities/player/particles.py
import pygame
import math
import random

# Import the original BloodParticle class instead of reimplementing it
from entities.blood_particle import BloodParticle

class ParticleSystem:
    """Manages XP and blood particles for the player"""
    def __init__(self, player):
        self.player = player
        self.xp_particles = []
        self.blood_particles = []
        self.max_blood_particles = 15
        
        # Map of stuck particles (organized by block coordinates)
        # Format: {(block_x, block_y): [(x, y, size, color, life), ...]}
        self.stuck_particles = {}
        self.current_block = (0, 0)
    
    def create_xp_particles(self, amount):
        """Create particles when XP is gained"""
        # Create 5-10 particles per XP point (capped for performance)
        particle_count = min(20, amount * 5)
        
        # Player center coordinates
        center_x = self.player.x + self.player.width / 2
        center_y = self.player.y + self.player.height / 2
        
        for _ in range(particle_count):
            # Random position around player
            angle = random.uniform(0, 2 * math.pi)
            distance = random.uniform(10, 30)
            
            pos_x = center_x + math.cos(angle) * distance
            pos_y = center_y + math.sin(angle) * distance
            
            # Random size
            size = random.randint(2, 4)
            
            # Golden color for XP particles
            self.xp_particles.append({
                'x': pos_x,
                'y': pos_y,
                'size': size,
                'color': (255, 215, 0),  # Gold color
                'life': random.randint(10, 20)
            })
    
    def spawn_blood_particles(self, damage_amount):
        """Create blood particles based on damage amount"""
        # Clear old particles
        self.blood_particles = []
        
        # Calculate particle count based on damage (more damage = more blood)
        particle_count = min(self.max_blood_particles, damage_amount * 3)
        
        # Player center coordinates
        center_x = self.player.x + self.player.width / 2
        center_y = self.player.y + self.player.height / 2
        
        # Create particles in a circular spray
        for _ in range(particle_count):
            # Random angle for particle direction
            angle = random.uniform(0, 2 * math.pi)
            
            # Direction vector from angle
            dir_x = math.cos(angle)
            dir_y = math.sin(angle)
            
            # Create particle - use the original BloodParticle class
            particle = BloodParticle(center_x, center_y, dir_x, dir_y)
            self.blood_particles.append(particle)
    
    def spawn_enemy_blood(self, enemy):
        """Spawn blood particles from an enemy when hit with improved positioning"""
        # Player center coordinates (source of the hit)
        player_center_x = self.player.x + self.player.width / 2
        player_center_y = self.player.y + self.player.height / 2
        
        # Enemy center coordinates
        enemy_center_x = enemy.x + enemy.width / 2
        enemy_center_y = enemy.y + enemy.height / 2
        
        # Direction vector from player to enemy
        dir_x = enemy_center_x - player_center_x
        dir_y = enemy_center_y - player_center_y
        
        # Normalize direction
        length = max(0.1, math.sqrt(dir_x**2 + dir_y**2))
        dir_x /= length
        dir_y /= length
        
        # Calculate impact point (on the edge of the enemy facing the player)
        impact_offset = min(enemy.width, enemy.height) * 0.4  # 40% of enemy size
        impact_x = enemy_center_x - dir_x * impact_offset
        impact_y = enemy_center_y - dir_y * impact_offset
        
        # Create particles
        particle_count = random.randint(8, 12)
        
        # Initialize blood_particles list for enemy if needed
        if not hasattr(enemy, 'blood_particles'):
            enemy.blood_particles = []
        
        # Create particles with direction biased away from player
        for _ in range(particle_count):
            # Random angle in a cone facing away from player
            base_angle = math.atan2(dir_y, dir_x)
            angle_variation = random.uniform(-0.7, 0.7)  # +/- 40 degrees
            angle = base_angle + angle_variation
            
            # Final direction vector with some randomness
            particle_dir_x = math.cos(angle) 
            particle_dir_y = math.sin(angle)
            
            # Add a bit of randomness to particle starting position
            rand_offset = 3.0
            start_x = impact_x + random.uniform(-rand_offset, rand_offset)
            start_y = impact_y + random.uniform(-rand_offset, rand_offset)
            
            # Larger particles for enemies - use original BloodParticle
            size = random.randint(4, 8)
            
            # Create particle
            particle = BloodParticle(start_x, start_y, particle_dir_x, particle_dir_y, size)
            
            # Add to enemy's blood particles
            enemy.blood_particles.append(particle)
    
    def update(self, current_time, obstacles=None):
        """Update all particles"""
        # Update XP particles
        for particle in self.xp_particles[:]:
            # Move particle if it has velocity
            if 'vel_x' in particle and 'vel_y' in particle:
                particle['x'] += particle['vel_x']
                particle['y'] += particle['vel_y']
                
                # Optional: Add gravity effect for some particles
                # particle['vel_y'] += 0.05  # Uncomment for gravity
                
                # Optional: Slow down particles over time
                particle['vel_x'] *= 0.95  # Add air resistance
                particle['vel_y'] *= 0.95
            
            # Decrease life
            particle['life'] -= 1
            
            # Remove dead particles
            if particle['life'] <= 0:
                self.xp_particles.remove(particle)
        
        # Update blood particles
        if obstacles:
            # Update each particle
            for particle in self.blood_particles[:]:
                particle.update(obstacles)
                
                # Remove inactive particles
                if not particle.active:
                    self.blood_particles.remove(particle)
                
                # If particle is stuck, add it to the stuck particles list for current block
                if hasattr(particle, 'stuck') and particle.stuck and particle in self.blood_particles:
                    self.blood_particles.remove(particle)
                    
                    # Make sure the current block has an entry in the dictionary
                    if self.current_block not in self.stuck_particles:
                        self.stuck_particles[self.current_block] = []
                    
                    # Store information about the stuck particle
                    self.stuck_particles[self.current_block].append((
                        particle.x,
                        particle.y,
                        particle.size,
                        particle.color,
                        particle.current_life
                    ))
    
    def draw_active_blood(self, surface):
        """Draw only the active (flying) blood particles"""
        # Draw active particles
        for particle in self.blood_particles:
            particle.draw(surface)

    def draw_stuck_blood(self, surface):
        """Draw only the stuck blood particles (should be called before drawing entities)"""
        # Draw stuck particles for the current block only
        if self.current_block in self.stuck_particles:
            for x, y, size, color, life in self.stuck_particles[self.current_block]:
                # Skip if outside screen (optimization)
                if x < -size or x > pygame.display.get_surface().get_width() + size or \
                y < -size or y > pygame.display.get_surface().get_height() + size:
                    continue
                
                # Calculate opacity based on life
                opacity = min(255, life * 10)  # Fade out based on life
                
                # Create a surface with alpha for the particle
                particle_surface = pygame.Surface((size, size), pygame.SRCALPHA)
                
                # Set the color with opacity
                color_with_alpha = (*color, opacity)
                pygame.draw.circle(particle_surface, color_with_alpha, 
                                (size//2, size//2), size//2)
                
                # Draw the particle
                surface.blit(particle_surface, (x - size//2, y - size//2))
    
    def draw_xp_particles(self, surface):
        """Draw XP particles with fading effect"""
        if not self.xp_particles:
            return
            
        for particle in self.xp_particles:
            # Calculate opacity based on remaining life percentage
            if 'max_life' in particle:
                opacity = int(255 * (particle['life'] / particle['max_life']))
            else:
                opacity = 255  # Full opacity for particles without max_life
            
            # Create a surface with alpha for the particle
            particle_surface = pygame.Surface((particle['size']*2, particle['size']*2), pygame.SRCALPHA)
            
            # Set the color with opacity
            color_with_alpha = (*particle['color'], opacity)
            pygame.draw.circle(particle_surface, color_with_alpha, 
                          (particle['size'], particle['size']), 
                          particle['size'])
            
            # Draw the particle
            surface.blit(particle_surface, (int(particle['x'] - particle['size']), int(particle['y'] - particle['size'])))
    
    def set_current_block(self, block_x, block_y):
        """Update the player's current block coordinates"""
        self.current_block = (block_x, block_y)
        
        # Initialize empty list for this block if it doesn't exist
        if self.current_block not in self.stuck_particles:
            self.stuck_particles[self.current_block] = []

    def create_fire_trail(self, x, y):
        # Small orange-red particles behind the projectile
        for _ in range(2):
            self.xp_particles.append({
                'x': x,
                'y': y,
                'size': random.randint(2, 4),
                'color': (255, random.randint(50, 150), 0),
                'life': random.randint(5, 15)
            })

    def create_fire_explosion(self, x, y):
        """Create a dynamic fire explosion with particles that spread outward"""
        # Create more particles for a bigger explosion
        particle_count = random.randint(15, 25)
        
        # Calculate explosion center
        center_x = x
        center_y = y
        
        # Create various fiery particles
        for _ in range(particle_count):
            # Random angle for 360-degree spread
            angle = random.uniform(0, 2 * math.pi)
            
            # Direction vector from angle
            dir_x = math.cos(angle)
            dir_y = math.sin(angle)
            
            # Random distance from center for starting position
            distance = random.uniform(0, 8)
            
            # Starting position with some randomness
            start_x = center_x + dir_x * distance
            start_y = center_y + dir_y * distance
            
            # Random size for fire particles
            size = random.randint(3, 8)
            
            # Fire colors: from bright yellow to deep orange/red
            r = 255
            g = random.randint(50, 200)  # More variation in green
            b = random.randint(0, 50)    # Small amount of blue for some particles
            
            # Velocity based on direction with some randomization
            velocity_mult = random.uniform(0.8, 2.5)
            
            # Create fire particle with velocity and fading
            self.xp_particles.append({
                'x': start_x,
                'y': start_y,
                'vel_x': dir_x * velocity_mult,  # Add velocity for movement
                'vel_y': dir_y * velocity_mult,
                'size': size,
                'color': (r, g, b),
                'life': random.randint(10, 25),
                'max_life': 25  # Track max life for fading
            })
        
        # Add a few embers/sparks (small, bright particles)
        for _ in range(5):
            angle = random.uniform(0, 2 * math.pi)
            dir_x = math.cos(angle)
            dir_y = math.sin(angle)
            
            # Embers move faster and further
            velocity_mult = random.uniform(2.0, 3.5)
            
            self.xp_particles.append({
                'x': center_x,
                'y': center_y,
                'vel_x': dir_x * velocity_mult,
                'vel_y': dir_y * velocity_mult,
                'size': random.randint(1, 3),  # Smaller particles
                'color': (255, 255, random.randint(100, 200)),  # Brighter, yellowish
                'life': random.randint(5, 15),
                'max_life': 15
            })
        
        # Add central flash effect
        self.xp_particles.append({
            'x': center_x,
            'y': center_y,
            'size': random.randint(10, 15),
            'color': (255, 255, 200),  # Bright white-yellow
            'life': random.randint(3, 8),
            'max_life': 8,
            'vel_x': 0,
            'vel_y': 0
        })

    def create_smoke_cloud(self, x, y):
        """Create a smoke cloud that rises and dissipates"""
        # Small rising black/gray cloud after firebolt hits
        for _ in range(8):
            # Random angle but biased upward
            angle = random.uniform(-2.5, 0.5)  # Mostly upward (-π to π/6)
            
            # Direction vector from angle
            dir_x = math.cos(angle) * 0.3  # Slower horizontal movement
            dir_y = math.sin(angle) * 0.5  # Faster vertical movement
            
            # Random starting position
            offset_x = random.uniform(-10, 10)
            offset_y = random.uniform(-5, 5)
            
            # Random gray color
            gray = random.randint(30, 80)
            
            # Create smoke particle
            self.xp_particles.append({
                'x': x + offset_x,
                'y': y + offset_y,
                'vel_x': dir_x,
                'vel_y': dir_y,
                'size': random.randint(4, 8),
                'color': (gray, gray, gray),
                'life': random.randint(20, 35),
                'max_life': 35
            })

File: web/src/utils/particles.js
/**
 * Particle System - Manages all game particles (blood, fire, XP, etc.)
 * Mirrors Python entities/player/particles.py
 */
class ParticleSystem {
    constructor() {
        this.particles = [];
        this.bloodParticles = [];
        this.maxParticles = 500;
    }

    /**
     * Create fire trail particles behind projectile
     */
    createFireTrail(x, y, size = null, life = null) {
        if (size === null) size = 2 + Math.random() * 2;
        if (life === null) life = 3 + Math.floor(Math.random() * 4);

        for (let i = 0; i < 2; i++) {
            this.particles.push({
                x: x,
                y: y,
                vx: 0,
                vy: 0,
                size: size,
                color: { r: 255, g: 50 + Math.floor(Math.random() * 100), b: 0 },
                life: life,
                maxLife: life,
                type: 'fire'
            });
        }
    }

    /**
     * Create fire explosion at impact point
     */
    createFireExplosion(x, y) {
        const particleCount = 15 + Math.floor(Math.random() * 11);

        // Main fire particles
        for (let i = 0; i < particleCount; i++) {
            const angle = Math.random() * Math.PI * 2;
            const dirX = Math.cos(angle);
            const dirY = Math.sin(angle);
            const distance = Math.random() * 8;

            const velocityMult = 0.8 + Math.random() * 1.7;

            this.particles.push({
                x: x + dirX * distance,
                y: y + dirY * distance,
                vx: dirX * velocityMult,
                vy: dirY * velocityMult,
                size: 3 + Math.random() * 5,
                color: { r: 255, g: 50 + Math.floor(Math.random() * 150), b: Math.floor(Math.random() * 50) },
                life: 10 + Math.floor(Math.random() * 15),
                maxLife: 25,
                type: 'fire'
            });
        }

        // Embers/sparks
        for (let i = 0; i < 5; i++) {
            const angle = Math.random() * Math.PI * 2;
            const velocityMult = 2.0 + Math.random() * 1.5;

            this.particles.push({
                x: x,
                y: y,
                vx: Math.cos(angle) * velocityMult,
                vy: Math.sin(angle) * velocityMult,
                size: 1 + Math.random() * 2,
                color: { r: 255, g: 255, b: 100 + Math.floor(Math.random() * 100) },
                life: 5 + Math.floor(Math.random() * 10),
                maxLife: 15,
                type: 'ember'
            });
        }

        // Central flash
        this.particles.push({
            x: x,
            y: y,
            vx: 0,
            vy: 0,
            size: 10 + Math.random() * 5,
            color: { r: 255, g: 255, b: 200 },
            life: 3 + Math.floor(Math.random() * 5),
            maxLife: 8,
            type: 'flash'
        });
    }

    /**
     * Spawn blood particles from enemy when hit
     */
    spawnEnemyBlood(enemy, playerX, playerY) {
        // Direction from player to enemy
        const enemyCenterX = enemy.x + enemy.width / 2;
        const enemyCenterY = enemy.y + enemy.height / 2;

        let dirX = enemyCenterX - playerX;
        let dirY = enemyCenterY - playerY;

        // Normalize
        const length = Math.max(0.1, Math.sqrt(dirX * dirX + dirY * dirY));
        dirX /= length;
        dirY /= length;

        // Impact point on enemy facing player
        const impactOffset = Math.min(enemy.width, enemy.height) * 0.4;
        const impactX = enemyCenterX - dirX * impactOffset;
        const impactY = enemyCenterY - dirY * impactOffset;

        // Create particles
        const particleCount = 8 + Math.floor(Math.random() * 5);

        for (let i = 0; i < particleCount; i++) {
            // Angle biased away from player
            const baseAngle = Math.atan2(dirY, dirX);
            const angleVariation = (Math.random() - 0.5) * 1.4; // +/- 40 degrees
            const angle = baseAngle + angleVariation;

            const particleDirX = Math.cos(angle);
            const particleDirY = Math.sin(angle);

            // Random offset from impact point
            const randOffset = 3.0;
            const startX = impactX + (Math.random() - 0.5) * randOffset * 2;
            const startY = impactY + (Math.random() - 0.5) * randOffset * 2;

            const speed = 1.0 + Math.random() * 3.0;
            const size = 4 + Math.floor(Math.random() * 5);

            this.bloodParticles.push({
                x: startX,
                y: startY,
                vx: particleDirX * speed,
                vy: particleDirY * speed,
                size: size,
                color: { r: 120 + Math.floor(Math.random() * 100), g: Math.floor(Math.random() * 30), b: Math.floor(Math.random() * 30) },
                life: 15 + Math.floor(Math.random() * 15),
                maxLife: 30,
                gravity: 0.2,
                friction: 0.95,
                stuck: false
            });
        }
    }

    /**
     * Create XP gain particles
     */
    createXpParticles(x, y, amount) {
        const particleCount = Math.min(20, amount * 5);

        for (let i = 0; i < particleCount; i++) {
            const angle = Math.random() * Math.PI * 2;
            const distance = 10 + Math.random() * 20;

            this.particles.push({
                x: x + Math.cos(angle) * distance,
                y: y + Math.sin(angle) * distance,
                vx: 0,
                vy: -0.5 - Math.random(),
                size: 2 + Math.floor(Math.random() * 3),
                color: { r: 255, g: 215, b: 0 },
                life: 10 + Math.floor(Math.random() * 10),
                maxLife: 20,
                type: 'xp'
            });
        }
    }

    /**
     * Update all particles
     */
    update(obstacles = []) {
        // Update fire/effect particles
        this.particles = this.particles.filter(p => {
            // Move particle
            p.x += p.vx;
            p.y += p.vy;

            // Apply friction to velocity
            p.vx *= 0.95;
            p.vy *= 0.95;

            // Decrease life
            p.life--;

            // Shrink over time
            if (p.type === 'fire' || p.type === 'ember') {
                p.size *= 0.97;
            }

            return p.life > 0 && p.size > 0.5;
        });

        // Update blood particles
        this.bloodParticles = this.bloodParticles.filter(p => {
            if (p.stuck) {
                p.life--;
                return p.life > 0;
            }

            // Apply gravity
            p.vy += p.gravity;

            // Apply friction
            p.vx *= p.friction;
            p.vy *= p.friction;

            // Move
            p.x += p.vx;
            p.y += p.vy;

            // Check collision with obstacles
            for (const obs of obstacles) {
                if (obs.getRect) {
                    const rect = obs.getRect();
                    if (p.x >= rect.x && p.x <= rect.x + rect.width &&
                        p.y >= rect.y && p.y <= rect.y + rect.height) {
                        p.stuck = true;
                        break;
                    }
                }
            }

            p.life--;
            return p.life > 0;
        });

        // Cap particles
        if (this.particles.length > this.maxParticles) {
            this.particles = this.particles.slice(-this.maxParticles);
        }
        if (this.bloodParticles.length > this.maxParticles) {
            this.bloodParticles = this.bloodParticles.slice(-this.maxParticles);
        }
    }

    /**
     * Render all particles
     */
    render(ctx) {
        // Draw fire/effect particles
        for (const p of this.particles) {
            const alpha = Math.min(1, p.life / p.maxLife);
            ctx.globalAlpha = alpha;
            ctx.fillStyle = `rgb(${p.color.r}, ${p.color.g}, ${p.color.b})`;
            ctx.beginPath();
            ctx.arc(p.x, p.y, p.size, 0, Math.PI * 2);
            ctx.fill();
        }

        // Draw blood particles
        for (const p of this.bloodParticles) {
            const fadeStart = p.maxLife / 2;
            let alpha = 1;
            if (p.life < fadeStart) {
                alpha = p.life / fadeStart;
            }
            ctx.globalAlpha = alpha;
            ctx.fillStyle = `rgb(${p.color.r}, ${p.color.g}, ${p.color.b})`;
            ctx.beginPath();
            ctx.arc(p.x, p.y, p.size / 2, 0, Math.PI * 2);
            ctx.fill();
        }

        ctx.globalAlpha = 1;
    }

    /**
     * Clear all particles (e.g., on block transition)
     */
    clear() {
        this.particles = [];
        this.bloodParticles = [];
    }
}

// Global instance
window.particleSystem = new ParticleSystem();
